include/nbl/builtin/hlsl/bxdf/brdf/specular/common.hlsl

```diff
@@ -0,0 +1,46 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_BXDF_BRDF_SPECULAR_COMMON_INCLUDED_
+#define _NBL_BUILTIN_HLSL_BXDF_BRDF_SPECULAR_COMMON_INCLUDED_
+
+#include <nbl/builtin/hlsl/bxdf/common.hlsl>
+#include <nbl/builtin/hlsl/bxdf/ndf/common.hlsl>
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace bxdf
+{
+namespace brdf
+{
+namespace specular
+{
+
+template <class fresnel_t, class ndf_t, class Sample, class Interaction, class MicrofacetCache>
+struct CookTorrance : BxDFBase<float3, float, Sample, Interaction, MicrofacetCache>
+{
+    fresnel_t fresnel;
+    ndf::ndf_traits<ndf_t> ndf;
+};
+
+template <class IncomingRayDirInfo, class fresnel_t, class ndf_t>
+struct IsotropicCookTorrance : CookTorrance<fresnel_t, ndf_t, LightSample<IncomingRayDirInfo>, surface_interactions::Isotropic<IncomingRayDirInfo>, IsotropicMicrofacetCache>
+{
+
+};
+
+template <class IncomingRayDirInfo, class fresnel_t, class ndf_t>
+struct AnisotropicCookTorrance : CookTorrance<fresnel_t, ndf_t, LightSample<IncomingRayDirInfo>, surface_interactions::Anisotropic<IncomingRayDirInfo>, AnisotropicMicrofacetCache>
+{
+
+};
```

Comment:
Its a bit sad that HLSL202x doesn't have SFINAE, because we could have just made isotropic and anisotropic the same struct and `enable_if`-ed the anisotropic interaction overload

However we have inheritance, so we can make the base class Isotropic, and inherit to make the Anistropic and add that overload, not as clean but will do

Reply:
but why? this completely messes up typedefs got from BxDFBase

Reply:
hmm, good observation.

The thing is that an anisotropic BxDF should always be usable as an isotropic one (thanks to inheritance).

Maybe the BxDFBase should be adjusted (sepearate typedefs for aniso and iso)?